F12 Associating a model to an Asset Type

JayChouteau opened the pull request:
This pull request integrates the feature F10, which allows the user to choose which RUL model to choose for a specific Asset Type. Here are the main points of the PR:
0) SQL scripts
- Make sure the latest dump file has been run (v13)
- Run the model_evaluation script on top of that (in folder v14)
1) Try feature
To try out the feature, go to the model tab in Asset Type info view. From there, you will be able to scroll horizontally through the list of models, see their RMSE value, select the model by clicking on the box as well as saving the association by clicking on save. Saving the feature will update the association in the database as well as set the trained_model to be retrained and the assets associated with that asset type to be updated.
2) Evaluation of models
This pull request also includes the evaluation of single models and not necessarily all of them. A button will be available for each model. Clicking on the button will generate the evaluation and update the corresponding RMSE value/values
3) Input Validation
- Models cannot be evaluated as long as the train slider value or test slider value is 0
- Save button cannot be clicked if no model is selected.
4) Verification of the feature
When saving, the model_id should be properly set in the trained_model table for the corresponding asset type. The retrain tag should also be set to true. Finally, the updated tag of all live assets of that asset type should be set to true.
5) UI
The UI looks relatively bad. This was left on purpose to be done by the UI Refactoring that is currently performed by Jeff and avoid duplicate work.
6) REFACTORING
there are 2 refactoring commits in this pull request. These refactoring types are "extract methods"

JayChouteau committed:
Creation of the model boxes (#218)
- Creation of a model object
- Creation of the model UI boxes

JayChouteau committed:
Merge branch '#216_Model_Evaluation' of https://github.com/rdsaliba/notorious-eng into F10-Association_AType-Model

 Conflicts:
	JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java
	JavaFXUserInterface/src/main/java/external/ModelDAOImpl.java

JayChouteau committed:
Updating DB with selected model (#218)
- Changing model ID in trained model
- Changing the retrain tag in trained model
- changing the updated tag for all assets of that asset type

JayChouteau committed:
Changes to the selection and generation of models (#218)
- selection of panes
- select button to generate button

JayChouteau committed:
Merging develop (#218)
- resolving conflicts with F10

JayChouteau committed:
Refactoring F10-F12(#218)
- Naming
- Comments
- Extraction of methods
-

JayChouteau committed:
REFACTORING + Evaluation classes (#218)
- REFACTORING: Extraction of methods
- Adding single model evaluation methods

JayChouteau committed:
REFACTORING + Evaluation classes (#218)
- Input validation on the sliders
- Displaying RMSE Value
- Evaluation function for a single model vs all models
- Scrolling through the models
- Model Selection

JayChouteau committed:
Removing of Cancel Button (#218)

JayChouteau committed:
Updating information (#218)
- updating RMSE value after model evaluation
- updating associated model information on info tab

JayChouteau committed:
REFACTORING (#218)
- Adding break statements to the switch statement in evaluation to increase performance and code convention

JayChouteau committed:
Enabling Evaluation buttons (#218)

JayChouteau committed:
Code convention enforcement (#218)
- Code cleanup and formatting
- Comments

TalalBazerbachi reviewed (changes requested):
when you evaluate models a new list is added instead of just updating the old one.
<img width="930" alt="image" src="https://user-images.githubusercontent.com/55160650/109230445-93680180-77cd-11eb-8eb5-2ff2e8c43fcd.png">

TalalBazerbachi reviewed (changes requested):
Saving a model for an asset type is not working properly. 
I tried selecting "Linear" and saving it then I went to another page.
When I went back to the asset type page, the model wasn't selected.

TalalBazerbachi reviewed (changes requested):
Last sprint we have agreed to add unit tests for each issue.
So can you add unit tests for the functions that you created?

JayChouteau committed:
Keeping track on the selected model (#218)
- Adjustments from review comments
- Method commenting

JayChouteau commented:
@t98b Here are the answers to your comments:
1 - for updating the thumbnails properly without creating new thumbnails, I left it as is as we agreed in our discussion. This error will be dealt with in your next pull-request that tackles the evaluation of models/UI relation. 
2 - I pushed a commit that solves the issue of keeping track of the current associated model on the page. The associated model is highlighted when the tab is loaded.
3 - I didn't do any unit tests for my methods. We did agree to add them to every issue, so I will be adding them shortly.

Shirwarsame commented on JavaFXUserInterface/src/main/java/Controllers/AssetTypeInfoController.java:
```diff
@@ -495,4 +555,158 @@ public boolean formInputValidation() {
+    /**
+     * Saves the association between the selected model and the current asset type in the database.
+     * This function also set the retrain tag to true to indicate that the model chosen for the asset
+     * type needs to be retrain. Finally, it labels all live assets of that asset type as updated.
+     *
+     * @author Jeremie
+     */
+    private void saveSelectedModelAssociation() {
+        modelDAO.updateModelAssociatedWithAssetType(selectedModel.getModelID(), assetType.getId());
+        modelDAO.setModelToTrain(assetType.getId());
+        updateAssetsForSelectedModelAssociation(assetType.getId());
+    }
```
The database is not updated when the save button is clicked once another model is selected.

JayChouteau replied:
@Shirwarsame @t98b The reason why the selection was not saved is because the save button could not be clicked in the first place. The calls attached to the save button were good, but the button could never be clicked because there was another UI element on top of it that would prevent the clicking. The element was a transparent anchor pane. I removed it in my latest commit. Sorry for that. Also, clicking on the save button now takes you back to the asset type list. When you click again on the type you modified, the modifications should appear.

Shirwarsame reviewed (changes requested):
General issues brought by Talal seem to have been answered. Missing unit tests and new issue found

JayChouteau committed:
Review comments (#218)
- Enabling save button
- Added unit tests

JayChouteau commented:
@t98b @Shirwarsame I added unit tests for the model java file. I am currently adding unit tests for the rest of my code.

JayChouteau committed:
Unit tests and class extraction (#218)
- Class extraction -> ModelPanes
- Unit tests for ModelPanes

JayChouteau committed:
Unit tests and DAO adjustments (#218)
- Model DAO unit tests
- Asset DAO unit tests
- Signature change for some DAO methods

JayChouteau commented:
@t98b @Shirwarsame I added unit tests for several methods implemented. I also extracted some methods of the assettypeinfocontroller into a separate class that I called ModelPanes. This was done to try to preserve the SOLID principles and avoid a Blob Class. Can you take a look at it and give your comments. Thanks!

JayChouteau committed:
SonarQube Refactoring (#218)

Shirwarsame reviewed (approved):
Excellent work the unit tests work and the changes that you made addressed the issues that were brought up
